icu4c/source/common/ulocale.cpp

```diff
@@ -0,0 +1,107 @@
+// © 2023 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+//
+#include "unicode/errorcode.h"
+#include "unicode/stringpiece.h"
+#include "unicode/utypes.h"
+#include "unicode/ustring.h"
+#include "unicode/ulocale.h"
+#include "unicode/locid.h"
+
+#include "cmemory.h"
+
+U_NAMESPACE_USE
+#define EXTERNAL(i) ((ULocale*)(i))
+#define CONST_INTERNAL(e) ((const icu::Locale*)(e))
+#define INTERNAL(e) ((icu::Locale*)(e))
+
+ULocale*
+ulocale_openForLocaleID(const char* localeID, int32_t length) {
+    if (length < 0 || localeID[length] == '\0') {
+        return EXTERNAL(icu::Locale::createFromName(localeID).clone());
+    } else {
+        if (length < ULOC_FULLNAME_CAPACITY-1) {
+            char buf[ULOC_FULLNAME_CAPACITY];
+            uprv_memcpy(buf, localeID, length);
+            buf[length] = '\0';
+            return EXTERNAL(icu::Locale::createFromName(buf).clone());
+        }
+        // return a bogus one
+        Locale bogus;
+        bogus.setToBogus();
+        return EXTERNAL(bogus.clone());
+    }
+}
+
+ULocale*
+ulocale_openForLanguageTag(const char* tag, int32_t length, UErrorCode* err) {
+  if (U_FAILURE(*err)) return nullptr;
```

Comment:
Won't `Locale::forLanguageTag()` check `U_FAILURE()` for you?  Or is the issue the call to `clone()` at the end?